### PR TITLE
Update user agent

### PR DESCRIPTION
**Pull request by WilliamBergamin**
###  Summary

The goal of this PR is to provide slack with a more meaningful `user-agent` information, when users make http requests to Slack.

Previously Deno would populate this header with `Deno/1.x.x` describing the users Deno version, now it will be populated with `Deno/1.x.x OS/<user os> deno-slack-api/<version | undefined>`

- [user os list](https://deno.land/api?s=Deno.build)

#### Testing

The unit tests try to cover all edge cases

Manual testing is tricky since the module must be released in order to get a successful version.

To test an undefined api version use the following to start a local server
```ts
const server = Deno.listen({ port: 8080 });
console.log(`HTTP webserver running.  Access it at:  http://localhost:8080/`);

for await (const conn of server) {
  serveHttp(conn);
}

async function serveHttp(conn: Deno.Conn) {

  const httpConn = Deno.serveHttp(conn);

  for await (const requestEvent of httpConn) {
	  console.log(requestEvent.request.headers.get("user-agent"));
    requestEvent.respondWith(
      new Response(`{}`, {
        status: 200,
      }),
    );
  }
}
```

Then use the following to send a request
```ts
import { SlackAPI } from "https://raw.githubusercontent.com/slackapi/deno-slack-api/update-user-agent/src/mod.ts"

const client = SlackAPI("test-token",
{
	slackApiUrl: "http://localhost:8080/"
});

await client.api.test();
```

The command line should print out something like `Deno/1.36.0 OS/darwin deno-slack-api/undefined`

### Requirements

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/deno-slack-api/blob/main/.github/CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

### CODE REVIEW

File: src/api_test.ts [hunk 1]
```diff
@@ -366,41 +368,6 @@ Deno.test("SlackAPI class", async (t) => {
   mf.uninstall();
 });
 
-Deno.test("serializeData helper function", async (t) => {
```
Comment (WilliamBergamin) [inline src/api_test.ts:369]: These tests where moves to `base-client-helpers_test.ts`

File: src/base-client.ts [hunk 2]
```diff
@@ -87,23 +90,3 @@ export class BaseSlackAPIClient implements BaseSlackClient {
     };
   }
 }
-
-// Serialize an object into a string so as to be compatible with x-www-form-urlencoded payloads
-export function serializeData(data: Record<string, unknown>): URLSearchParams {
```
Comment (WilliamBergamin) [inline src/base-client.ts:92]: This function was moved to `base-client-helpers.ts`

File: src/base-client-helpers_test.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      const getModuleUrlStub = stub(_internals, "getModuleUrl", () => {
+        return new URL("https://deno.land/x/deno_slack_api@2.1.0/mod.ts)");
+      });
+
+      try {
+        const moduleVersion = _internals.getModuleVersion();
+
+        assertSpyCalls(getModuleUrlStub, 1);
+        assertEquals(moduleVersion, "2.1.0");
+      } finally {
```
Comment (filmaj) [inline src/base-client-helpers_test.ts:19]: For my own education, why do we wrap the assertions in try/finally? What would happen here if an exception is raised in `getModuleVersion()`?

File: src/base-client-helpers_test.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        getModuleUrlStub.restore();
+      }
+    },
+  );
+
+  await t.step(
+    "should return the unknown if the module is not sourced from deno.land",
```
Comment (filmaj) [inline src/base-client-helpers_test.ts:26]:
```suggestion
    "should return undefined if the module is not sourced from deno.land",
```

File: src/base-client-helpers_test.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        assertSpyCalls(getModuleUrlStub, 1);
+        assertEquals(moduleVersion, undefined);
+      } finally {
+        getModuleUrlStub.restore();
+      }
+    },
+  );
+
+  await t.step(
+    "should return the unknown if the module version is invalid",
```
Comment (filmaj) [inline src/base-client-helpers_test.ts:43]: What's the idea in this test? How is the version invalid in the test case?

File: src/base-client-helpers_test.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
+import { assertEquals } from "https://deno.land/std@0.185.0/testing/asserts.ts";
+import { _internals, getUserAgent } from "./base-client-helpers.ts";
+import { serializeData } from "./base-client-helpers.ts";
+import { assertSpyCalls, stub } from "./dev_deps.ts";
+
+Deno.test(_internals.getModuleVersion.name, async (t) => {
```
Comment (filmaj) [inline src/base-client-helpers_test.ts:6]: Can you prefix all the test names with the module name it exists in, i.e. `base-client-helpers`? This way if the test fails in the future while we are working on something else, it is a bit more immediately obvious what failed and from where.

File: src/base-client-helpers_test.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      } finally {
+        getModuleUrlStub.restore();
+      }
+    },
+  );
+});
+
+Deno.test(getUserAgent.name, async (t) => {
+  await t.step(
+    "should return the user agent with expected output",
```
Comment (filmaj) [inline src/base-client-helpers_test.ts:62]: I'm a nitpicker when it comes to test names, but they are important. A test name can convey a lot of information about a program's behaviour without ever having to look at code. From the perspective of a new contributor unfamiliar with the project, good test names make it easier to ramp up, making it easier to contribute.

Let's be clear what 'expected output' means - you as the developer of this feature know what that means, but me or someone else does not. I would suggest instead:

```suggestion
    "should return the user agent with deno version, OS name and undefined deno-slack-api version",
```

**Review by filmaj — CHANGES_REQUESTED** [review]
Hey, looking good! Thanks for tackling this.

I left a bunch of nitpicky comments. mostly around the tests, and some questions for my own education.

One other thing we should try to look into: what does this look like for a deployed application? I'm not exactly sure how this works in the context of a bundled application. I think it would be useful to run the `build` hook on an app that uses a versioned deno-slack-api from deno.land and see how the code from this repo ends up getting bundled into the built bundle (if it does at all!), and try to reason through how this code would work in that context.

File: src/base-client-helpers_test.ts (hunk 3 above)
Comment (WilliamBergamin) [reply src/base-client-helpers_test.ts:19]: From what I can understand the resources get subbed on a global level and must be explicitly released :shrug:  an error is raised if another test tries to stub a stubbed resource

The `try/finally` pattern insures that the resource gets released with `getModuleUrlStub.restore();`, I've also thought of using [bdd](https://deno.land/manual@v1.36.3/basics/testing/behavior_driven_development) but opted for this pattern since it followed existing testing patterns

File: src/base-client-helpers_test.ts (hunk 4 above)
Comment (WilliamBergamin) [reply src/base-client-helpers_test.ts:26]: Good catch 💯

File: src/base-client-helpers_test.ts (hunk 5 above)
Comment (WilliamBergamin) [reply src/base-client-helpers_test.ts:43]: Good point, I will update the test description, this test validates that the Regex used to extract the version works properly and would return `undefined` if the module url path does not match `deno_slack_api@`

Example `deno_slack_sdk`

File: src/base-client-helpers_test.ts (hunk 6 above)
Comment (WilliamBergamin) [reply src/base-client-helpers_test.ts]: I've added the `base-client-helpers` module but I've hard coded it since I could not find a way to get this  value dynamically let me know if you know a better way to do this?

File: src/base-client-helpers_test.ts (hunk 7 above)
Comment (WilliamBergamin) [reply src/base-client-helpers_test.ts:65]: Great point, I did overlook the test names here 💯

File: src/base-client-helpers_test.ts (hunk 6 above)
Comment (filmaj) [reply src/base-client-helpers_test.ts]: That's fine!

**Comment by WilliamBergamin** [comment]
> what does this look like for a deployed application? I'm not exactly sure how this works in the context of a bundled application. I think it would be useful to run the build hook on an app that uses a versioned deno-slack-api from deno.land and see how the code from this repo ends up getting bundled into the built bundle (if it does at all!), and try to reason through how this code would work in that context.

### App testing

I've tested this by
1. Creating a new app
2. Importing the deno-slack-api from https://raw.githubusercontent.com/slackapi/deno-slack-api/update-user-agent/src/
3. Editing the `deno-slack-sdk` locally to import the `deno-slack-api` from https://raw.githubusercontent.com/slackapi/deno-slack-api/update-user-agent/src/
4. Importing this local `deno-slack-sdk` in my new app
5. `slack run` works as expected 🟢 
6. `slack deploy` works as expected 🟢 

### Remote module testing

I've release the [deno_bill_api](https://deno.land/x/deno_bill_api) and used the following script with the above server to get the value of the `user-agent`
```ts
import { SlackAPI } from "https://deno.land/x/deno_bill_api@0.0.1/mod.ts"

const client = SlackAPI("test-token",
{
	slackApiUrl: "http://localhost:8080/"
});

await client.api.test();
```

My locally deployed server received the following user agent
`user-agent = Deno/1.36.0 OS/darwin deno-slack-api/0.0.1` 🟢 

*Note* that deno-slack-api is printed instead of deno-bill-api, this is a typo in the deno-bill-api

### Bundle testing

I then used the `deno bundle` command on the above script to generate a bundled version of my script (`fetch.bundle.js`) when I execute this script (`deno run --allow-net fetch.bundle.js`) the `user-agent` received by my local server is `Deno/1.36.0 OS/darwin deno-slack-api/0.0.1` 🟢 

Turns out in the `fetch.bundle.js` file the following `importMeta` object stores import.url information of the modules 🥇 
```js
const importMeta = {
    url: "https://deno.land/x/deno_bill_api@0.0.1/base-client-helpers.ts",
    main: false
};
const API_VERSION_REGEX = /\/deno_bill_api@(.*)\//;
function getUserAgent() {
    const userAgents = [];
    userAgents.push(`Deno/${Deno.version.deno}`);
    userAgents.push(`OS/${Deno.build.os}`);
    userAgents.push(`deno-slack-api/${_internals.getModuleVersion()}`);
    return userAgents.join(" ");
}
function getModuleVersion() {
    const url = _internals.getModuleUrl();
    if (url.host === "deno.land") {
        return url.pathname.match(API_VERSION_REGEX)?.at(1);
    }
    return undefined;
}
function getModuleUrl() {
    return new URL(importMeta.url);
}
```

It may be fair to assume that this behavior will also apply to users using the `deno-slack-api` through the `deno-slack-sdk`

@filmaj let me know what you think  about this and if it is worth testing with an app that imports a version of `deno-slack-sdk` that references the `deno-bill-api`

**Comment by filmaj** [comment]
Oh man thanks for that extensive testing and deep dive! Very cool info about the `import.meta` bundling, that's great to know!

**Review by filmaj — APPROVED** [review]
Looks awesome! Well done and thank you for tackling this 🙇 ❤️ 🎉 

:shipit:
